### PR TITLE
Implement disconnect command

### PR DESCRIPTION
**Pull request by GiasemiSh**
closes https://github.com/QMSTR/qmstr-all/issues/194

Disconnects 

- Filenode from FileNode 

- Filenode from PackageNode

Needs https://github.com/QMSTR/qmstr-proto/pull/12 to work

### CODE REVIEW

File: pkg/cli/disconnect.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+var disconnectCmd = &cobra.Command{
+	Use:   "disconnect [type_of_node:attribute:value] [type_of_node:attribute:value]",
+	Short: "Disconnect nodes with specific edges",
+	Long: `Usage: qmstrctl disconnect <that> <this>...
+Disconnect from Node <that> Node(s) <this>.
+	
+input: [type_of_node:attribute:value]
+
+type_of_node:
```
Comment (hemarkus) [inline pkg/cli/disconnect.go]: Generate this information

File: pkg/cli/disconnect.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/QMSTR/qmstr/pkg/service"
+	"github.com/spf13/cobra"
+	"golang.org/x/net/context"
+)
+
+var disconnectCmdFlags = struct {
```
Comment (hemarkus) [inline pkg/cli/disconnect.go]: good, we should do this in other commands as well.

File: pkg/cli/disconnect.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		}
+		switch thisVal := thisID.(type) {
+		// FileNode -> FileNode
+		case *service.FileNode:
+			this, err := getUniqueFileNode(thisVal)
+			if err != nil {
+				return fmt.Errorf("get unique file node fail. please use better matching params: %v", err)
+			}
+			// default edge
+			if disconnectCmdFlags.edge == "" {
```
Comment (hemarkus) [inline pkg/cli/disconnect.go]: So why do you assign an empty string in the first place?

File: pkg/cli/disconnect.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		if err != nil {
+			return fmt.Errorf("get package node fail: %v", err)
+		}
+	default:
+		return fmt.Errorf("unsuported node type %T", thatVal)
+	}
+	return nil
+}
+
+func disconnectFromFileNode(node *service.FileNode, args []string) error {
```
Comment (hemarkus) [inline pkg/cli/disconnect.go]: What is args?

File: pkg/cli/disconnect.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		that, err := getUniqueFileNode(thatVal)
+		if err != nil {
+			return fmt.Errorf("get unique \"that\" node fail: %v", err)
+		}
+		err = disconnectFromFileNode(that, args[1:])
+		if err != nil {
+			return fmt.Errorf("disconnect file nodes fail: %v", err)
+		}
+	case *service.PackageNode:
+		_, err := controlServiceClient.GetPackageNode(context.Background(), &service.PackageNode{})
```
Comment (hemarkus) [inline pkg/cli/disconnect.go]: So you are checking if you can get the package node. What for? How will this disconnect anything? Please explain.

File: pkg/cli/disconnect.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			this, err := getUniqueFileNode(thisVal)
+			if err != nil {
+				return fmt.Errorf("get unique file node fail. please use better matching params: %v", err)
+			}
+			// default edge
+			if disconnectCmdFlags.edge == "" {
+				disconnectCmdFlags.edge = "derivedFrom"
+			}
+			// Which edge
+			switch disconnectCmdFlags.edge {
```
Comment (hemarkus) [inline pkg/cli/disconnect.go]: switch inside a case is rather bad style. Please clean this up.

File: pkg/cli/disconnect.go (hunk 3 above)
Comment (GiasemiSh) [reply pkg/cli/disconnect.go]: Default edge for a file node is 'derivedFrom' but default edge for a package node is 'targets'. So we can't have a defined default value.

File: pkg/cli/disconnect.go (hunk 4 above)
Comment (GiasemiSh) [reply pkg/cli/disconnect.go]: What do you mean?

File: pkg/cli/disconnect.go (hunk 5 above)
Comment (GiasemiSh) [reply pkg/cli/disconnect.go]: The package node case is implemented in a later commit

File: pkg/cli/disconnect.go (hunk 6 above)
Comment (GiasemiSh) [reply pkg/cli/disconnect.go]: done

**Comment by GiasemiSh** [comment]
@hemarkus please review again

File: pkg/cli/connect.go [hunk 7]
```diff
@@ -74,7 +73,7 @@ func connectCmdRun(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
-func connectToFileNode(node *service.FileNode, args []string) error {
+func connectToFileNode(that *service.FileNode, args []string) error {
 	for _, nID := range args {
```
Comment (hemarkus) [inline pkg/cli/connect.go]: - What is args? 
- Parsing nodes from the args is duplicated in the connectToPackageNode function.
- When parsing fails there might be nodes added (no transaction)

Proposed solution:
Parse all args then descend into the the connect functions, that have get passed a proper slice of nodes (or interface{})

**Comment by GiasemiSh** [comment]
And again.. @hemarkus Please review again

**Pull request merged**